libyul/tools/interpreter/PureEVMInstructionInterpreter.cpp

```diff
@@ -0,0 +1,312 @@
+	case Instruction::SELFDESTRUCT:
+		return ImpureBuiltinEncountered();
+
+	case Instruction::POP:
+		return EvaluationOk();
+	// --------------- invalid in strict assembly ---------------
+	case Instruction::JUMP:
+	case Instruction::JUMPI:
+	case Instruction::JUMPDEST:
+	case Instruction::PUSH0:
+	case Instruction::PUSH1:
+	case Instruction::PUSH2:
+	case Instruction::PUSH3:
+	case Instruction::PUSH4:
+	case Instruction::PUSH5:
+	case Instruction::PUSH6:
+	case Instruction::PUSH7:
+	case Instruction::PUSH8:
+	case Instruction::PUSH9:
+	case Instruction::PUSH10:
+	case Instruction::PUSH11:
+	case Instruction::PUSH12:
+	case Instruction::PUSH13:
+	case Instruction::PUSH14:
+	case Instruction::PUSH15:
+	case Instruction::PUSH16:
+	case Instruction::PUSH17:
+	case Instruction::PUSH18:
+	case Instruction::PUSH19:
+	case Instruction::PUSH20:
+	case Instruction::PUSH21:
+	case Instruction::PUSH22:
+	case Instruction::PUSH23:
+	case Instruction::PUSH24:
+	case Instruction::PUSH25:
+	case Instruction::PUSH26:
+	case Instruction::PUSH27:
+	case Instruction::PUSH28:
+	case Instruction::PUSH29:
+	case Instruction::PUSH30:
+	case Instruction::PUSH31:
+	case Instruction::PUSH32:
+	case Instruction::DUP1:
+	case Instruction::DUP2:
+	case Instruction::DUP3:
+	case Instruction::DUP4:
+	case Instruction::DUP5:
+	case Instruction::DUP6:
+	case Instruction::DUP7:
+	case Instruction::DUP8:
+	case Instruction::DUP9:
+	case Instruction::DUP10:
+	case Instruction::DUP11:
+	case Instruction::DUP12:
+	case Instruction::DUP13:
+	case Instruction::DUP14:
+	case Instruction::DUP15:
+	case Instruction::DUP16:
+	case Instruction::SWAP1:
+	case Instruction::SWAP2:
+	case Instruction::SWAP3:
+	case Instruction::SWAP4:
+	case Instruction::SWAP5:
+	case Instruction::SWAP6:
+	case Instruction::SWAP7:
+	case Instruction::SWAP8:
+	case Instruction::SWAP9:
+	case Instruction::SWAP10:
+	case Instruction::SWAP11:
+	case Instruction::SWAP12:
+	case Instruction::SWAP13:
+	case Instruction::SWAP14:
+	case Instruction::SWAP15:
+	case Instruction::SWAP16:
+	{
+		yulAssert(false, "");
+		return EvaluationOk(0);
+	}
+	}
+
+	yulAssert(false, "Unknown instruction with opcode " + std::to_string(static_cast<uint8_t>(_instruction)));
+	return EvaluationOk(0);
+}
+
+EvaluationResult PureEVMInstructionInterpreter::evalBuiltin(
+	BuiltinFunctionForEVM const& _fun,
+	std::vector<Expression> const& /* _arguments */,  // This was required to execute some builtin.
+													  // But all of them are impure.
+	std::vector<u256> const& _evaluatedArguments
+)
+{
+	if (_fun.instruction)
+		return eval(*_fun.instruction, _evaluatedArguments);
+
+	std::string fun = _fun.name.str();
+	bool isVerbatim = boost::starts_with(fun, "verbatim");
+	if (isVerbatim)
+		return ImpureBuiltinEncountered();
+
+	static std::set<std::string> const NON_INSTRUCTION_BUILTIN_NAME = {
+		"datasize",
+		"dataoffset",
+		"datacopy",
+		"memoryguard",
+		"loadimmutable",
+		"setimmutable",
+		"linkersymbol"
+	};
```

Comment:
Instead of hard-coding those, you should just check that the `instruction` field is empty.

Reply:
I will change as suggested. I was just afraid that there might be builtin other than instructions that can be evaluated in the future. This part was added to have the same runtime warning effect as I was doing with the switch exhaustive check.

Reply:
If there is, we'll quickly find out in tests and fix it, so don't worry about it.

The only other thing are IIRC the dynamically generated verbatim functions, but you reject those as well.

Reply:
This got fixed, along side making the `PureEVMInstructionInterpreter` a function.